### PR TITLE
Refactor PlaybackSerialiser

### LINKED ISSUE #164: Add subtick ability to keyboard (Similar to mouse paths)
I was made aware that the keyaboard can perform subtick inputs as well, similar to the mouse.

This is only really useful in guiscreens when using the keyboard to switch hotbar positions.

This also fixes the issue of having to hold certain keys for an input to be registered in low tickrates for example.

- [x] ~~Increase tasfile version to 2~~ Make TASFile versions strings.
- [x] Add unit tests to Virtual Input

## New syntax
After thinking about this for a while, I should probably introduce a new syntax for proper subtick handling

### Current
```
55|Keyboard:W,LCONTROL;w|Mouse:;[MOUSEMOVED,0,887,626]->[RC,0,1580,658]->[MOUSEMOVED,0,1580,658]|Camera:17.85;-202.74799~&
```
Subticks are stored in square brackets. (`[MOUSEMOVED,0,887,626]->[RC,0,1580,658]` etc...). This looks just very messy

### Proposal
```
55|Keyboard:W,LCONTROL;w|Mouse:;0,887,626|Camera:17.85;-202.74799
	1|Keyboard:|Mouse:RC;0,1580,658|Camera:17.85;-202.74799
	2|Keyboard:|Mouse:;0,1580,658|Camera:17.85;-202.74799
```
The non indented lines are for ticks, while the indented lines are for subticks.

This way, we can support Mouse and Keyboard subticks, as well as Camera subticks while we are at it.  
This also gets rid of the square bracket syntax.  
We can also use the existing comparing system for ticks to also check for subticks and maybe add monitoring and control bytes for subtick commands.
And this system is upgradable, the old syntax can be converted into the new one.

### LINKED ISSUE #183: Remove ~& at the end of datalines
This was to separate data from the monitoring comments, but I did this so crudely, surely there has to be a better way

### LINKED ISSUE #184: Add option for relative variables to TASfile
While we are at it with TASfile changes, let's swap pitch and yaw to make it similar to the minecraft f3 screen.

- [ ] Swap pitch and yaw in TASfile
- [ ] Add option to use relative variables for pitch and yaw: `~20;~-14`
- [ ] Do deltas for other absolute coordinates as well (Pointer location)
- [ ] Change scroll wheel to always have a ~ in front as that is always a delta. Allow for absolute variables

### LINKED ISSUE #193: Refactor PlaybackSerialiser
- [x] Move version specific code to it's own class
- [x] Create interface for the must haves

### PR DESCRIPTION
**Pull request by ScribbleTAS**
Part 3 of my ongoing quest to refactor the mod.

While the first PR (#179) was hooking into the vanilla code,  
the second PR (#200) was about storing the inputs and metadata in RAM.

In this PR I will tackle the task of storing all the inputs we recorded to the file.

# Current
The current system is (as always) a hardcoded mess of stuff somewhat randomly put together...  
Now with a bit more experience under my belt, I should be able to upgrade whatever I wrote 3 years ago...

# Changes
## PlaybackSerialiser
The serialiser has been completely rewritten to support the new subtick format that was introduced in #179.

### Alpha
```
0|Keyboard:;|Mouse:;[MOUSEMOVED,0,0,0]|Camera:0.0;0.0~&				//Monitoring:0.0 0.0 0.0 0.0 0.0 0.0 0
1|Keyboard:; |Mouse:;[MOUSEMOVED,0,-121,58]|Camera:0.0;0.0~&				//Monitoring:-38.874069623529465 56.0 1.4322641845001531 0.0 -0.0784000015258789 0.0 0
2|Keyboard:;|Mouse:;[MOUSEMOVED,0,-121,58]|Camera:0.0;0.0~&				//Monitoring:-38.874069623529465 56.0 1.4322641845001531 0.0 -0.0784000015258789 0.0 0
3|Keyboard:;|Mouse:;[MOUSEMOVED,0,-121,58]|Camera:0.0;0.0~&				//Monitoring:-38.874069623529465 56.0 1.4322641845001531 0.0 -0.0784000015258789 0.0 0
```
In Alpha, every line contained one tick, with the mouse being the only peripheral that supported subticks (in square brackets).  
Monitoring was put at the end of the line in a comment, with an infuriating ~& inbetween, which was the only way for me at the time to seperate data from comments.

Overall the file format looks very messy.

### Beta
```
23|||21.749935;-318.6463		// $desyncMonitor(-33.52014, 56.00000, -4.20817, 0.00000, -0.07840, 0.00000);
24|||		// $desyncMonitor(-33.52014, 56.00000, -4.20817, 0.00000, -0.07840, 0.00000);
25|||		// $desyncMonitor(-33.52014, 56.00000, -4.20817, 0.00000, -0.07840, 0.00000);
26|||		// $desyncMonitor(-33.52014, 56.00000, -4.20817, 0.00000, -0.07840, 0.00000);
27|||21.749935;-320.8963		// $desyncMonitor(-33.52014, 56.00000, -4.20817, 0.00000, -0.07840, 0.00000);
28|||21.149935;-322.5463		// $desyncMonitor(-33.52014, 56.00000, -4.20817, 0.00000, -0.07840, 0.00000);
	1|||20.849936;-322.84628
	2|||20.249935;-323.4463
	3|||19.499935;-324.34628
	4|||19.199936;-324.49628
	5|||19.049936;-324.64627
	6|||18.899937;-324.49628
29|LCONTROL;||18.749937;-324.1963		// $desyncMonitor(-33.59629, 56.00000, -4.10603, -0.04158, -0.07840, 0.05577);
	1|W,LCONTROL;^W||18.599937;-324.0463
	2|||18.599937;-323.8963
	3|||18.449938;-323.8963
	4|||18.449938;-323.7463
	5|||18.299938;-323.7463
	6|||18.299938;-323.5963
	7|||18.299938;-323.44632
	8|||18.149939;-323.44632
	9|||18.149939;-323.29633
```
Now in the beta, subticks are denoted with a TAB at the beginning, while the start of the tick is without it.
Keyboard, Mouse and Camera indicators have been removed as well as the ~& at the end of the file.

Monitoring is now done via "File Commands", which can be turned off if necessary.

### Relative input
For cursor and camera angle, it is now possible to put ~ in front of your value (e.g. ~45), instead of hard values.  
This will change the value relative to the previous value (e.g Add 45 units).

## API
This PR adds 3 new registries related to serialisation.

### PlaybackMetadata
Enables you to store things in the header of the TASfile
```java
public class YourMetadataExtension extends PlaybackMetadataExtension {

	@Override
	public String getExtensionName() {
		return "Your Extension";
	}

	@Override
	public PlaybackMetadata onStore() {
		PlaybackMetadata metadata = new PlaybackMetadata(this);
		metadata.setValue("key", "value");
		return metadata;
	}

	@Override
	public void onLoad(PlaybackMetadata metadata) {
		System.out.println(metadata.getValue("key"));
	}

	@Override
	public void onClear() {
		// Clear any stored things
	}

}
```
Register this with `TASmodAPIRegistry.PLAYBACK_METADATA.register(new YourMetadataExtension())`

The resulting header in the TASfile should contain something similar to this:
```
--------------------- Your Extension --------------------
key:value
```
### PlaybackFileCommands
File commands are additions to the TASfile in the form of `$name(args);` for extended functionality.  

In Alpha, this was done using a system named "control bytes" made by @PancakeTAS.  
It was possible to add certain lines to the TASfile, which executed functionality when the playback ran over that tick.
- `$interpolation on/off`: Turns on/off interpolation for playback
- `$hud on/off`: Turns on/off hud during playback
- `$info off/String`: Adds a box to the top left with a custom string

This functionality has now been moved to file commands:
- ~~interpolation~~: Removed, as interpolation is now handled by subticks, you can delete the camera angle subticks to selectively turn off interpolation.
- `$hud(true|false);`: Same as before
- `$label(String)`: To turn off, leave the String empty.

Additionally, desyncMonitoring was implemented during Alpha versions by being hardcoded at the end of the lines.  
As seen in the examples above, desyncMonitor is now a file command.

File commands have the benefit, that they are easily extendible and can be turned off if needed.

Other specifications:
- Multiple file commands can be in the same line: `//$hud(false); $info();`
- Comments can be in the same line as file commands: `//$hud(true); This is a comment`[^1]
- There are 2 types of file commands (and comments), Inline and Endline
  - Inline comments are comments before a tick line.  
  File commands in these comments are only performed on the whole tick.
  	```
  	// This is an inline comment $hud(false);
  	29|LCONTROL;||18.749937;-324.1963
	1|W,LCONTROL;^W||18.599937;-324.0463
	2|||18.599937;-323.8963
  	```
  - Endline comments appear at the end of a line. File commands are executed on a subtick level
    ```
  	29|LCONTROL;||18.749937;-324.1963	// This is an endline comment $desyncMonitor(etc..);
	1|W,LCONTROL;^W||18.599937;-324.0463
	2|||18.599937;-323.8963
    ```

Similar to MetadataExtensions, you can create your own by extending `PlaybackFileCommandExtension` and registering it with `TASmodAPIRegistry.PLAYBACK_FILE_COMMAND.register(new YourFileCommandExtension())`.

Flavors can be turned off with `/filecommand <fcname>`, while `/filecommand` shows, which fcs are enabled.

### SerialiserFlavor
If you want TASmod to support an entire new TASfile syntax, then you can create a new flavor[^2] by extending `SerialiserFlavorBase` and registering it with `TASmodAPIRegistry.SERIALISER_FLAVOR.register(new YourFlavorExtension)`;

You can specify the flavor during the `/saveTAS` command: `/saveTAS <name> [flavor]`. When specifying a flavor, it will be set as the default, so the second save will not need the flavor.

The defaultFlavor is currently called "beta1".

## Renames
- mctcommon/server => mctcommon/networking
- Moved TASmodKeybinds to registries package
- Moved TASmodPackets to registries package
- Split EventClient and EventServer into more files:
   - EventTickrate
   - EventSavestate
   - EventVirtualInput
## Fixes
- Fixed F8,F9 not being recognised in chat
- Fixed TASmodConfig not being extendible

[^1]: File commands can appear after or inbetween a comment, but when saving, file commands will be moved to the beginning of the file.

[^2]: It's called flavors, as markup languages like Markdown are not entirely the same across websites and apps. These versions are called "flavors". For example the github flavor of Markdown supports footnotes, while other flavors do not.

# TODO
## CustomSerialisers
- [x] Create a serialiser registry, so it's easy to create your own TASfile Syntax
- [x] Create a base class that all serialisers need to extend from
- [x] Create new serialiser that supports the new VirtualInputFormat
	- [x] #183
	- [x] #184

## PlaybackExtensions
- [x] Add PlaybackExtensionRegistry
- [x] Convert DesyncMonitor to extension
- [x] Convert ControlBytes to extension

### CODE REVIEW

**Comment by ScribbleTAS** [comment]
Note to self: 
Do new registry stuff

**Comment by ScribbleTAS** [comment]
Note to self:
Continue doing registry stuff!
Then make a new one for flavors and start writing the new system!

**Comment by ScribbleTAS** [comment]
Note to self:
Continue on PlaybackFlavorBase and BetaFlavor. Get unit tests up and running for flavors to be able to test them!

**Comment by ScribbleTAS** [comment]
Note to self:
Deserialiser time!

**Comment by ScribbleTAS** [comment]
Alright, code seems to be finished but everything is totally untested..

So now all thats left is writing tests, error handling and documentation... And writing the actual file format but that is quite easy at this point...

Note to self:
- [x] Test mouse deserialisation
- [x] Test camera deserialisation
- [x] Test isNumeric
- [x] Test isFloat
- [x] Test deserialiseContainer
	- [x] Add extracting comments and extracting tick
	- [x] Test splitContainer
- [x] Add test to PlaybackSerialiser
- [x] Add error handling
- [x] Test stuff with error handling

**Comment by PancakeTAS** [comment]
Read to merge? /j

**Comment by ScribbleTAS** [comment]
Note to self:  
Remove filecommand parameter nonsense from FlavorBase...
Move all handling to the FileCommandRegistry. Enable file command extensions through that system statically instead passing the list of enabled stuff through the Flavor...

Serialise everything in serialiseInline and serialiseEndline comments. Let's finish this damn module!

**Comment by ScribbleTAS** [comment]
Alright, I finally have a somewhat good grip on how to handle filecommands and how to combine 2 APIs together in a good way...

I have now offloaded the handling code to the FileCommandRegistry as that part should not be editable via the API.  
Messy serialising code as I somehow have to convert a 2 dimensional list like this:

```
List {
	List{
		$desyncMonitor(); 	// A command handled by the desyncMonitorExtension
		$desyncMonitor();	// A second command in the next subtick handled by the desyncMonitorExtension
		$desyncMonitor();
		$desyncMonitor();
		$desyncMonitor();
		$desyncMonitor();
	}
	List{
		$interpolation(true); 	// A command handled by the interpolationExtension
		$interpolation(false); 	// A second command in the next subtick
		$interpolation(true);
		$interpolation(true);
		$interpolation(true);
		$interpolation(true);
	}
}
```
which is ideal for storing the commands, when each extension stores all of their PlaybackFileCommands themselves
to something like this:

```
List{
	List{					// This is now a list of PlaybackCommands 
		$desyncMonitor();
		$interpolation(true);
	}
	List{
		$desyncMonitor();
		$interpolation(false);
	}
	List{
		$desyncMonitor();
		$interpolation(true);
	}
	// etc...
}
```
which is ideal for serialisation as one List now contains all FileCommands for one subtick...
Now the first subtick can easily be serialised to `// $desyncMonitor(); $interpolation(true); A comment!`.

Best part is, that I have to do the reverse now for deserialisation...

So, note to self:
- Implement `handleOnDeserialisation` in PlaybackRegistry
- Fix Disabled tests
- Test the s*** out of everything

**Comment by ScribbleTAS** [comment]
Note to self:
Allow null in PlaybackFileCommandContainer

**Comment by ScribbleTAS** [comment]
Note to self:
> Allow null in PlaybackFileCommandContainer
Nevermind, fixed it already...
- Test numerical keycodes being serialised and deserialised correctly
- Test and add ErrorHandling

**Comment by ScribbleTAS** [comment]
Note to self:
- [x] Test relative values in TASfile
- [x] Test error when flavor can not be found

**Comment by ScribbleTAS** [comment]
Final stretch!
Note to self:
- [x] Upgrade networking of `/save` and `/load` to account for flavors
- [x] Add default flavor on client for quicker saving
- [x] Hook up the serialiser with savestates

**Comment by ScribbleTAS** [comment]
Note to self:
- [x] Remove camera angle stuff in file
~~- [ ] Add K, M, C~~

**Comment by ScribbleTAS** [comment]
Note to self:
- [x] Add command for fileCommands
- [x] Add config option for file commands

**Comment by PancakeTAS** [comment]
:partying_face: :partying_face: :partying_face: :partying_face: :partying_face:
you'll be getting my full review in approx 3-5 business days

File: src/main/java/com/minecrafttas/mctcommon/Configuration.java [hunk 1]
```diff
@@ -5,40 +5,75 @@
...
  * A <i>very</i> simple configuration class
+ * 
  * @author Scribble
- *
  */
 
-public class Configuration {
-	
+public class Configuration extends AbstractRegistry<com.minecrafttas.mctcommon.Configuration.ConfigOptions> {
+
```
Comment (PancakeTAS) [inline src/main/java/com/minecrafttas/mctcommon/Configuration.java]: Why the long import string?

File: src/main/java/com/minecrafttas/mctcommon/Configuration.java [hunk 2]
```diff
@@ -5,40 +5,75 @@
...
+	
+	@Override
+	public void register(ConfigOptions registryObject) {
+		if(registryObject == null) {
+			return;
+		}
+		
+		if(configRegistry.contains(registryObject)) {
+			return;
+		}
```
Comment (PancakeTAS) [inline src/main/java/com/minecrafttas/mctcommon/Configuration.java:49]: i think there's a null-check inside of contains, but whatever

File: src/main/java/com/minecrafttas/mctcommon/Configuration.java [hunk 3]
```diff
@@ -5,40 +5,75 @@
...
+	public void load() {
+		if (file.exists()) {
+			properties = loadInner();
 		}
-		if(properties == null || !file.exists()) {
+		if (properties == null || !file.exists()) {
 			properties = generateDefault();
 			save();
-		}
+		}
```
Comment (PancakeTAS) [inline src/main/java/com/minecrafttas/mctcommon/Configuration.java:74]: can you please turn on your save actions to fix these whitespaces automatically, haha

File: src/main/java/com/minecrafttas/tasmod/TASmodClient.java [hunk 4]
```diff
@@ -255,7 +213,7 @@ public void onPlayerJoinedClientSide(EntityPlayerSP player) {
 		}
 		
 		
-		if(!(ip+":"+port).equals(connectedIP)) {
+		if(!(ip+":"+port).equals(connectedIP)) { // TODO Clean this up. Make TASmodNetworkHandler out of this... Maybe with Permission system?
 			try {
```
Comment (PancakeTAS) [inline src/main/java/com/minecrafttas/tasmod/TASmodClient.java:217]: TODO spotted

File: src/main/java/com/minecrafttas/tasmod/commands/TabCompletionUtils.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+		File[] listOfFiles = folder.listFiles(new FileFilter() {
+			@Override
+			public boolean accept(File pathname) {
+				return pathname.getName().endsWith(".mctas");
+			}
+		});
+		for (int i = 0; i < listOfFiles.length; i++) {
+			tab.add(listOfFiles[i].getName().replaceAll("\\.mctas", ""));
+		}
```
Comment (PancakeTAS) [inline src/main/java/com/minecrafttas/tasmod/commands/TabCompletionUtils.java]: There's .replace() for non-regex replace

**Review by PancakeTAS — APPROVED** [review]
few nitpicks, got only through half of the files before I got demotivated
image

File: src/main/java/com/minecrafttas/tasmod/playback/metadata/PlaybackMetadataRegistry.java [hunk 6]
```diff
@@ -3,73 +3,46 @@
...
+import com.minecrafttas.tasmod.playback.metadata.PlaybackMetadata.PlaybackMetadataExtension;
 
 /**
  * Registry for registering custom metadata that is stored in the TASFile.<br>
  * <br>
  * The default metadata includes general information such as author name,
  * savestate/rerecord count and category.<br>
  * <br>
- * Any custom class has to extend PlaybackMetadataExtension
+ * Any custom class has to implement PlaybackMetadataExtension
```
Comment (ScribbleTAS) [inline src/main/java/com/minecrafttas/tasmod/playback/metadata/PlaybackMetadataRegistry.java]: Javadoc is wrong, should be "extend"

File: src/main/java/com/minecrafttas/tasmod/playback/tasfile/exception/PlaybackLoadException.java [hunk 7]
```diff
@@ -1,12 +1,55 @@
...
+		this(cause, String.format(msg, args));
+	}
+	
+	private static String printTick(long line, long tick, int subtick) {
+		return String.format("Line %s, Tick %s, Subtick %s: ", line, tick, subtick);
+	}
+	
+	@Override
+	public void printStackTrace() {
+		TASmod.LOGGER.catching(this);;
```
Comment (ScribbleTAS) [inline src/main/java/com/minecrafttas/tasmod/playback/tasfile/exception/PlaybackLoadException.java]: Double ;

File: src/main/java/com/minecrafttas/mctcommon/Configuration.java (hunk 1 above)
Comment (ScribbleTAS) [reply src/main/java/com/minecrafttas/mctcommon/Configuration.java]: The compiler threw an error and eclipse just refused to import it... Kinda forgot about it

**Pull request merged**